### PR TITLE
feat: support changelog templates used during GitHub Release creation

### PR DESCRIPTION
**Pull request by KevinDeJong-TomTom**
You can configure the content of your GitHub Release using the `release.y[a]ml` configuration file stored in `.github`, for ex.
```yaml
changelog:
  exclude:
    labels:
      - dependencies
  categories:
    - title: ⚠️ Breaking Changes
      labels:
        - bump:major
    - title: 🚀 New Features
      labels:
        - bump:minor
    - title: 🐛 Bug Fixes
      labels:
        - bump:patch
    - title: 🚧 Other changes
      labels:
        - "*"
```

During generation, each individual conventional commit will be checked against the labels denoting the SemVer bump (`bump:<version>`) to determine its associated category.

> **NOTE**: The `bump:<version>` label set on your Pull Request will be ignored in favor of
individual commits

Please refer to the ["Automatically generated release notes"](https://docs.github.com/en/repositories/releasing-projects-on-github/automatically-generated-release-notes#configuring-automatically-generated-release-notes) documentation for more details

### CODE REVIEW

File: src/changelog.ts [hunk 1]
```diff
@@ -108,6 +119,77 @@ function getIssueReferenceSuffix(commit: ConventionalCommitMessage): string {
   return "";
 }
 
+/**
+ * Creates an entry in the Changelog based on the provided
+ * Conventional Commit message.
+ */
+async function generateChangelogEntry(
```
Comment (KevinDeJong-TomTom) [inline src/changelog.ts:126]: Note; only moved code to its own function

File: src/changelog.ts [hunk 2]
```diff
@@ -118,44 +200,87 @@ export async function generateChangelog(
...
+
+    // Check the individual commit Bump Level
+    exclusionPatterns.labels?.push(bumpLabel);
+
+    if (hasConfigurationElement(exclusionPatterns, config.changelog.exclude)) {
+      continue;
     }
 
-    config.get(commit.bump)?.changes.push(msg);
+    for (const category of config.changelog.categories) {
```
Comment (jstmlr) [inline src/changelog.ts:246]: This seems unnecessary to do on every commit, as the config data never changes.

Try getting rid of this for-loop, e.g. consider a bulking up the data from `config.changelog.categories` once and doing a single call to `hasConfigurationElement()`.

File: src/changelog.ts [hunk 3]
```diff
@@ -118,44 +200,87 @@ export async function generateChangelog(
...
+
+        // Check the labels in the Pull Request associated with this commit
+        for (const label of pullRequest.labels) {
+          // Ignore bump-labels on Pull Requests
+          if (label.name.startsWith("bump:")) {
+            continue;
+          }
+
+          exclusionPatterns.labels?.push(label.name);
+        }
```
Comment (jstmlr) [inline src/changelog.ts:231]: Consider something like:
```
  exclusionPatterns.labels = pullRequest.labels.filter(x => !x.name.startsWith("bump:")).map(x => x.name)
```

File: src/changelog.ts [hunk 4]
```diff
@@ -108,6 +119,77 @@ function getIssueReferenceSuffix(commit: ConventionalCommitMessage): string {
...
+    }
+  }
+
+  return JSON.parse(JSON.stringify(DEFAULT_CONFIG));
+}
+
+/**
+ * Checks whether the provided labels/authors are part of the reference set
+ */
+function hasConfigurationElement(
```
[the remark this answers is not in the document]
Comment (KevinDeJong-TomTom) [reply src/changelog.ts:170]: I suppose that I could replace this method completely with `.some()` statements, eg:

```ts
        if (pullRequest.user && config.changelog.exclude?.authors?.includes(pullRequest.user.login)) {
          continue;
        }
```

and...
```ts
    if (exclusionLabels.some(label => config.changelog.exclude?.labels?.includes(label))) {
      continue;
    }
```

and...

File: src/changelog.ts [hunk 5]
```diff
@@ -16,52 +16,63 @@
...
+ * Exclude pattern, part of the Release Configuration
  */
-interface IChangelogCategory {
-  /* Title message to use in the Changelog */
-  title: string;
-  /* Emoji to display in the Changelog */
-  emoji: string;
-  /* List of changes associated with this Changelog category */
-  changes: string[];
+interface IExcludeConfiguration {
```
[the remark this answers is not in the document]
Comment (KevinDeJong-TomTom) [reply src/changelog.ts:27]: Will no longer refer to this interface anywhere other than as part of the Configuration; will therefore keep the name as-is for now (see https://github.com/tomtom-international/commisery-action/pull/179#discussion_r1049714067)

File: src/changelog.ts (hunk 2 above)
Comment (KevinDeJong-TomTom) [reply src/changelog.ts:223]: As discussed - will make a separate task to improve error handling (i.e. find duplications, ..), which will also rely on remapping data entries.

**Review by jstmlr — APPROVED** [review]

**Pull request merged**